Add support for coalesced MMIO (`KVM_CAP_COALESCED_MMIO` /  `KVM_CAP_COALESCED_PIO`)

00xc opened the pull request:
### Summary of the PR

Add support for coalesced MMIO. This performance feature allows guest writes to port and memory space to not trigger VM exits. Instead, the kernel will write an entry into a shared ring buffer for each access, which userspace must consume. The ring buffer is mapped at a certain offset in the vcpu's file descriptor.

In order to enable this capability, introduce the `KvmCoalescedIoRing` struct, which will act as a safe wrapper around the raw mapping of the ring buffer. Since users may not want to use coalesced MMIO, or it might not be available, store it as an `Option` in the `VcpuFd` struct.

Add two tests as well for the newly added code.

The public API consists of:
* A couple of capabilities:
    - `Cap::CoalescedMmio` (`KVM_CAP_COALESCED_MMIO`)
    - `Cap::CoalescedPio` (`KVM_CAP_COALESCED_PIO`).
* Methods to register and unregister addresses that will use this feature:
    - `VmFd::register_coalesced_mmio()` (`KVM_REGISTER_COALESCED_MMIO` ioctl).
    - `VmFd::unregister_coalesced_mmio()` (`KVM_UNREGISTER_COALESCED_MMIO` ioctl).
* A shared ring buffer (mapped with `VcpuFd::map_coalesced_mmio_ring()`).
* A safe method to read from the ring (`VcpuFd::read_coalesced_mmio()`).

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [X] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [X] All added/changed functionality has a corresponding unit/integration
  test.
- [X] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [X] Any newly added `unsafe` code is properly documented.

00xc committed:
Add support for coalesced MMIO (KVM_CAP_COALESCED_MMIO)

Add support for coalesced MMIO. This performance feature allows guest
writes to port and memory space to not trigger VM exits. Instead, the
kernel will write an entry into a shared ring buffer for each access,
which userspace must consume. The ring buffer is mapped at a certain
offset in the vcpu's file descriptor.

In order to enable this capability, introduce the KvmCoalescedIoRing
struct, which will act as a safe wrapper around the raw mapping of
the ring buffer. Since users may not use coalesced MMIO, or it might
not be available, store it as an Option in the VcpuFd struct.

Signed-off-by: Carlos López <carlos.lopez@suse.com>

00xc committed:
Add tests for coalesced MMIO

Add tests for coalesced MMIO and PIO respectively. Both tests are
similar. They:

1. Set up a guest
2. Register a coalesced MMIO/PIO address
3. Map the ring buffer
4. Trigger a MMIO/PIO write
5. Check that the access did not trigger a VM exit
6. Check that there is a new matching entry in the ring buffer.
7. Unregister the MMIO/PIO address
8. Trigger a MMIO/PIO write
9. Check that the access triggered a VM exit

Signed-off-by: Carlos López <carlos.lopez@suse.com>

JonathanWoollett-Light commented on src/ioctls/mod.rs:
```diff
@@ -26,6 +29,100 @@ pub mod vm;
+            return Err(errno::Error::last());
+        }
+        Ok(Self {
+            addr: addr.cast::<kvm_coalesced_mmio_ring>(),
```
Nit: I think it can infer the type of this cast.
```suggestion
            addr: addr.cast(),
```

JonathanWoollett-Light commented on src/ioctls/mod.rs:
```diff
@@ -26,6 +29,100 @@ pub mod vm;
+        // SAFETY: This is safe because we mmap the page ourselves, and nobody
+        // else is holding a reference to it.
+        unsafe {
+            libc::munmap(self.addr.cast::<libc::c_void>(), self.page_size);
```
Nit: I think it can infer the type of this cast.
```suggestion
            libc::munmap(self.addr.cast(), self.page_size);
```

JonathanWoollett-Light commented on src/ioctls/vcpu.rs:
```diff
@@ -1743,6 +1745,56 @@ impl VcpuFd {
+        let ring = KvmCoalescedIoRing::mmap_from_fd(&self.vcpu)?;
+        self.coalesced_mmio_ring = Some(ring);
+        Ok(())
+    }
```
```suggestion
    pub fn map_coalesced_mmio_ring(&mut self) -> Result<()> {
        if self.coalesced_mmio_ring.is_none() {
            let ring = KvmCoalescedIoRing::mmap_from_fd(&self.vcpu)?;
            self.coalesced_mmio_ring = Some(ring);
        }
        Ok(())
    }
```

JonathanWoollett-Light commented on src/ioctls/vcpu.rs:
```diff
@@ -1743,6 +1745,56 @@ impl VcpuFd {
+            return Err(errno::Error::new(libc::EIO));
+        };
+        Ok(ring.read_entry())
+    }
```
```suggestion
    pub fn coalesced_mmio_read(&mut self) -> Result<Option<kvm_coalesced_mmio>> {
        let ring = self.coalesced_mmio_ring.ok_or(errno::Error::new(libc::EIO))?;
        Ok(ring.read_entry())
    }
```
You can also do
```rust
    pub fn coalesced_mmio_read(&mut self) -> Result<Option<kvm_coalesced_mmio>> {
        self.coalesced_mmio_ring
            .ok_or(errno::Error::new(libc::EIO))
            .map(|ring|ring.read_entry())
    }
```

JonathanWoollett-Light commented on src/ioctls/vcpu.rs:
```diff
@@ -2826,4 +2883,199 @@ mod tests {
+        const DATA: u64 = 0x39;
+        const SIZE: u32 = 1;
+
+        #[rustfmt::skip]
+        let code = [
+            0xe6, 0x2c,   // out 0x2c, al
+            0xf4,         // hlt
+            0xe6, 0x2c,   // out 0x2c, al
+            0xf4,         // hlt
+        ];
```
Can we use constants for the mnemonics?

E.g.
```suggestion
        use iced_x86::Mnemonic::{Hlt,Out};
        
        #[rustfmt::skip]
        let code = [
            Out as u8, 0x2c,
            Hlt as u8,
            Out as u8, 0x2c,
            Hlt as u8,
        ];
```

JonathanWoollett-Light commented on src/ioctls/vcpu.rs:
```diff
@@ -2826,4 +2883,199 @@ mod tests {
+        let mem_size = 0x4000;
+        let load_addr = mmap_anonymous(mem_size);
+        let guest_addr: u64 = 0x1000;
+        let slot: u32 = 0;
```
```suggestion
        let slot = 0;
```

JonathanWoollett-Light commented on src/ioctls/vcpu.rs:
```diff
@@ -2826,4 +2883,199 @@ mod tests {
+        const DATA: u64 = 0x39;
+        const SIZE: u32 = 2;
+
+        #[rustfmt::skip]
+        let code = [
+            0x66, 0x31, 0xFF,        // xor di,di
+            0x66, 0xBF, 0x24, 0x01,  // mov di, 0x124
+            0x67, 0x66, 0x89, 0x05,  // mov WORD PTR [di], ax
+            0xF4,                    // hlt
+            0x66, 0x31, 0xFF,        // xor di,di
+            0x66, 0xBF, 0x24, 0x01,  // mov di, 0x124
+            0x67, 0x66, 0x89, 0x05,  // mov WORD PTR [di], ax
+            0xF4,                    // hlt
+        ];
```
Can we use constants for the mnemonics?

E.g.
```suggestion
        use iced_x86::Mnemonic::{Xor, Mov, Hlt};
        
        #[rustfmt::skip]
        let code = [
            Xor as u8, 0x31, 0xFF,        // xor di,di
            Mov as u8, 0xBF, 0x24, 0x01,  // mov di, 0x124
            Mov as u8, 0x66, 0x89, 0x05,  // mov WORD PTR [di], ax
            Hlt as u8,                    // hlt
            Xor as u8, 0x31, 0xFF,        // xor di,di
            Mov as u8, 0xBF, 0x24, 0x01,  // mov di, 0x124
            Mov as u8, 0x66, 0x89, 0x05,  // mov WORD PTR [di], ax
            Hlt as u8,                    // hlt
        ];
```

00xc commented:
@JonathanWoollett-Light addressed your suggestions in the newer version, except the ones about using function mnemonics. Regarding that, isn't that change out of scope for this PR? It would add a new dependency and we would still have a bunch of tests which use raw bytes as assembly code. I'd rather work on that change in a separate PR.

00xc commented:
By the way the code can also properly be assembled with that crate's `CodeAssembler`, e.g.

```rust
let mut asm = CodeAssembler::new(16).unwrap();
asm.out(0x2c, al).unwrap();
asm.hlt().unwrap();
asm.out(0x2c, al).unwrap();
asm.hlt().unwrap();
let code = asm.assemble(base_addr).unwrap();
```

roypat commented on src/ioctls/mod.rs:
```diff
@@ -26,6 +29,100 @@ pub mod vm;
+    }
+
+    /// Gets a mutable reference to the ring
+    fn ring_mut(&mut self) -> &mut kvm_coalesced_mmio_ring {
+        // SAFETY: We have a `&mut self` and the pointer is private, so this
+        // access is exclusive.
+        unsafe { &mut *self.addr }
+    }
```
Just to be sure, since access to this only happens on a KVM_EXIT path, KVM will also not concurrently modify the mmap'd memory region while we hold this &mut?

00xc replied:
As far as I know KVM will not append entries to the ring buffer while the vCPU is stopped. That would make an implementation very difficult, even in C, as the ring buffer is not atomic.

roypat replied:
Sounds good, thanks for clarifying!

00xc replied:
Just to elaborate:
1. qemu seems to work under this assumption (see [`kvm_flush_coalesced_mmio_buffer()`](https://github.com/qemu/qemu/blob/8fa379170c2a12476021f5f50d6cf3f672e79e7b/accel/kvm/kvm-all.c#L2663))
2. This feature is not very well documented in KVM, but briefly looking at the kernel code, KVM simply avoids exiting to user space when it gets a VM exit with a write to an address registered for coalesced I/O. In other words, updates to the ring buffer do not happen on a vCPU that has exited to userspace, as there are no further VM exits.

roypat commented on src/ioctls/vcpu.rs:
```diff
@@ -2826,4 +2882,199 @@ mod tests {
+        let kvm = Kvm::new().unwrap();
+        let vm = kvm.create_vm().unwrap();
+        if !vm.check_extension(Cap::CoalescedPio) {
+            return;
```
Can we replace this `return` with a `panic`, to make sure we're actually running the test in CI?

00xc replied:
Fixed in the new version.

roypat commented on src/ioctls/vcpu.rs:
```diff
@@ -2826,4 +2882,199 @@ mod tests {
+        let kvm = Kvm::new().unwrap();
+        let vm = kvm.create_vm().unwrap();
+        if !vm.check_extension(Cap::CoalescedMmio) {
+            return;
```
Same here

00xc replied:
Fixed in the new version.

JonathanWoollett-Light commented:
> @JonathanWoollett-Light addressed your suggestions in the newer version, except the ones about using function mnemonics. Regarding that, isn't that change out of scope for this PR? It would add a new dependency and we would still have a bunch of tests which use raw bytes as assembly code. I'd rather work on that change in a separate PR.

Yeah this makes sense.

00xc commented:
Rebased on main and fixed conflicts

00xc commented:
Gentle ping for reviewers :)

JonathanWoollett-Light commented:
@00xc Can you please resolve the merge conflicts, after this LGTM.

00xc commented:
Rebased on main and fixed conflicts
